### ISSUE TITLE
The precision of triangle attention is inconsistent with PyTorch's standard precision.

### ISSUE THREAD

**Issue by lupengk** [NVIDIA/cuEquivariance#135]
Implementing efficient triangle attention is a great work !,,, but during testing, I noticed significant precision discrepancies compared to the PyTorch version. Could there be an issue with my testing approach? I noticed that a pytest script for the torch version of triangle attention hasn't been provided yet.

OUPUT is:
max value of cuequivariance tri-att: tensor(3.5000, device='cuda:0', dtype=torch.bfloat16, grad_fn=<MaxBackward1>) max value of reference tri-att: tensor(2.5469, device='cuda:0', dtype=torch.bfloat16, grad_fn=<MaxBackward1>) max diff is tensor(3.6875, device='cuda:0', dtype=torch.bfloat16, grad_fn=<MaxBackward1>)

CODE is:

import torch
import math
from cuequivariance_torch import triangle_attention

@torch.jit.ignore
def softmax_cast(t: torch.Tensor, dim: int = -1) -> torch.Tensor:
    d = t.dtype
    if d is torch.bfloat16:
        with torch.autocast("cuda", enabled=True):
            s = torch.nn.functional.softmax(t, dim=dim)
    else:
        s = torch.nn.functional.softmax(t, dim=dim)
    return s.to(d)

def reference_attention(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    biases: list,
) -> torch.Tensor:
    # [*, H, C_hidden, K]
    key = key.transpose(-1, -2)

    # [*, H, Q, K]
    a = torch.matmul(query, key)

    for b in biases:
        a += b

    a = softmax_cast(a, -1)

    # [*, H, Q, C_hidden]
    a = torch.matmul(a, value)

    return a

if torch.cuda.is_available():
    device = torch.device("cuda")
    # Set up dimensions
    batch_size, seq_len, seq_len2, num_heads, hidden_dim = 1, 280,570, 2, 32
    # Create input tensors on GPU with bfloat16 precision
    q = torch.randn(batch_size, seq_len, num_heads, seq_len, hidden_dim,
                    device=device, dtype=torch.bfloat16, requires_grad=True)
    k = torch.randn(batch_size, seq_len, num_heads, seq_len2, hidden_dim,
                    device=device, dtype=torch.bfloat16, requires_grad=True)
    v = torch.randn(batch_size, seq_len, num_heads, seq_len2, hidden_dim,
                    device=device, dtype=torch.bfloat16, requires_grad=True)
    bias = torch.randn(batch_size, 1, num_heads, seq_len, seq_len2,
                       device=device, dtype=torch.bfloat16, requires_grad=True)
    # Create optional mask
    mask = torch.rand(batch_size, seq_len, 1, 1, seq_len2,
                      device=device) < 0.5
    # Calculate scale
    scale = 1 / math.sqrt(hidden_dim)
    # Forward pass
    output, lse, max_val = triangle_attention(
        q=q, k=k, v=k, bias=bias, mask=mask, scale=scale, return_aux=True)
    output2 = reference_attention(q*scale,k,v, biases=[bias,(mask.float()-1)*1e9]  )
    print('max value of cuequivariance tri-att:', torch.max(torch.abs(output)),'max value of reference tri-att:',torch.max(torch.abs(output2)), 'max diff is',torch.max(torch.abs(output-output2)))

**Comment by yzhautouskay** [comment]
hey @lupengk, there is a typo in arguments: triangle_attention(q=q, k=k, **v=k**

**Comment by yzhautouskay** [comment]
yes, numerical differences are expected between cuEquivariance kernels and naive PyTorch reference. In end-to-end workloads, large magnitude values are unlikely due to normalization though, and relative error is what matters the most

**Comment by hsadasivan** [comment]
Marking as closed as we haven't heard back
